### PR TITLE
adapt screens to use PlacesWidget

### LINKED ISSUE #82: Update AddScreen to use PlacesWidget
When adding a trip, users should have a DropDownMenu with suggested locations, similar to the bootcamp.

### LINKED ISSUE #241: add test for checking duplicate creation of trips/activities
add tests for the bug fix implemented in the previous sprint - avoid duplicate creation of trips/activities by double clicking on save

### PR DESCRIPTION
**Pull request by mirceagavrizi**
## Summary

This PR implements the functionality of choosing a location through the PlacesSearchWidget for the add trip screen and add activity screen. Moreover, the tests have been adapted and the Location model as well.

## Changes

- **Models:** Objects of type Location have 5 parameters: id, name, address, lat, lng in order to have the details we need from a CustomPlace (we can not use directly the Place type as is not serializable)
- **Screens/UI:** Added a search bar with a clear button for the add activity screen and add trip screen.
- **Bug Fixes/Improvements:** Addresses #82 and #241 .
- **PlaceSearchWidget:** Added a function to convert a CustomPlace object to a Location object 

## Checklist

Ensure all / most of these are checked before the pull request is submitted.
- [x] This PR resolves #82 and #241 .
- [x] Tests have been added for new functionality.
- [x] Screenshots or UI changes have been attached (if applicable).
- [x] Documentation has been updated (if applicable).

##  Testing

- **Manual Testing:**
    - [x] Tested on emulator / physical device.
- **Unit Tests**
    - [x] Added units tests for this

##  Related Issues/Tickets

Closes #82 
CLoses #241 

## Screenshots (if applicable)

https://github.com/user-attachments/assets/3d3fe45c-08f3-4520-9ccd-2f4ac0c8cd3f

### CODE REVIEW

File: app/src/main/java/com/android/voyageur/ui/components/PlaceSearchWidget.kt [hunk 1]
```diff
@@ -20,22 +25,41 @@ import com.android.voyageur.model.place.PlacesViewModel
...
         modifier = modifier.fillMaxWidth().testTag("searchTextField"),
-        placeholder = { Text("Search places...") })
+        keyboardOptions = KeyboardOptions.Default.copy(imeAction = ImeAction.Search),
+        keyboardActions =
+            KeyboardActions(
+                onSearch = {
+                  val matchingPlace =
+                      searchedPlaces.firstOrNull { it.place.displayName == query.text }
+                  if (matchingPlace != null) {
+                    onSelect(matchingPlace)
```
Comment (apana03) [inline app/src/main/java/com/android/voyageur/ui/components/PlaceSearchWidget.kt:50]: We should remove this, we select the place only when the user clicks on an item from the dropdown, not when his query matches exactly the name of a given place

File: app/src/main/java/com/android/voyageur/ui/overview/AddTrip.kt [hunk 2]
```diff
@@ -187,18 +194,8 @@ fun AddTripScreen(
...
-                    Location(
-                        country = parts[0],
-                        city = parts[1],
-                        county = parts.getOrNull(2),
-                        zip = parts.getOrNull(3))
-                  } else {
-                    Location(country = "Unknown", city = "Unknown", county = null, zip = null)
-                  }
-                },
+                selectedPlace?.let { listOf(Location(address = it.place.displayName ?: "")) }
```
Comment (apana03) [inline app/src/main/java/com/android/voyageur/ui/overview/AddTrip.kt:197]: We should use the CustomPlace class to store the location, the Location class is deprecated as it was created at the beginning of the project. This causes loss of information since we do not only fetch the display name when searching for places (for example we also fetch the latitude and longitude which might be useful if we want to show different trips on map)

File: app/src/main/java/com/android/voyageur/ui/trip/AddActivity.kt [hunk 3]
```diff
@@ -208,7 +216,9 @@ fun AddActivityScreen(
         Activity(
             title = title,
             description = description,
-            location = Location(country = "Unknown", city = "Unknown", county = null, zip = null),
+            location =
+                selectedPlace?.let { Location(address = it.place.displayName ?: "") }
```
Comment (apana03) [inline app/src/main/java/com/android/voyageur/ui/trip/AddActivity.kt:220]: Same here, we should store the place as a Custom place not as a simple location.

File: app/src/main/java/com/android/voyageur/ui/components/PlaceSearchWidget.kt [hunk 4]
```diff
@@ -20,22 +25,41 @@ import com.android.voyageur.model.place.PlacesViewModel
 fun PlaceSearchWidget(
     placesViewModel: PlacesViewModel,
     onSelect: (CustomPlace) -> Unit,
-    modifier: Modifier = Modifier
+    modifier: Modifier = Modifier,
+    query: TextFieldValue,
+    onQueryChange: (TextFieldValue) -> Unit
```
Comment (apana03) [inline app/src/main/java/com/android/voyageur/ui/components/PlaceSearchWidget.kt:30]: Do we really need the query to be exported in the classes that use the search widget? I think the purpose of this widget was to be ready out of the box. Simply clicking on a dropdown item was going to perform the OnSelect function on that specific place.

Comment (apana03) [reply app/src/main/java/com/android/voyageur/ui/components/PlaceSearchWidget.kt:30]: However, we could leave it like this if it is absolutely necessary

**Review by apana03 — CHANGES_REQUESTED** [review]
Nice work! There are a few things that I believe need to be changed:

**Using the CustomPlace class**: We should use this new data class that was created in the last weeks, this is because this class contains the Google Place and a list of images for each location. It contains much more information than simply the display name, and this information is useful for the trip. Whilst I know that using the new location may entail cumbersome changes in the trip repository and activity repository I believe its use is essential

**Testing**: the usage of the PlacesView model is not thoroughly tested in the PR (only has about 60% line coverage), this has to be increased especially since adding a trip and an activity is one of the core functionalities of the app.

File: app/src/main/java/com/android/voyageur/ui/overview/AddTrip.kt (hunk 2 above)
Comment (mirceagavrizi) [reply app/src/main/java/com/android/voyageur/ui/overview/AddTrip.kt:206]: I adapted both Trip and Activity objects to use Place in order to store the location, as we don't need to save the photos in firebase. Let me know what do you think.

File: app/src/main/java/com/android/voyageur/ui/trip/AddActivity.kt (hunk 3 above)
Comment (mirceagavrizi) [reply app/src/main/java/com/android/voyageur/ui/trip/AddActivity.kt:220]: Same as above.

File: app/src/main/java/com/android/voyageur/ui/components/PlaceSearchWidget.kt (hunk 4 above)
Comment (mirceagavrizi) [reply app/src/main/java/com/android/voyageur/ui/components/PlaceSearchWidget.kt:31]: Just using the initial widget didn't work, so I had to add it as a parameter. Maybe for the next sprint we can find a workaround, but it is not a necessity.

**Comment by mirceagavrizi** [comment]
The line coverage for the PR is above 70% and in order for it to be above 80% we need to write tests for App.kt, which doesn't really make sense. Otherwise, I have added tests for all the features and edge cases for the other files. Also, this issue is blocking @g0g05arui, so it would be better if we merge this feature asap.

File: app/src/main/java/com/android/voyageur/ui/components/PlaceSearchWidget.kt [hunk 5]
```diff
@@ -46,11 +62,22 @@ fun PlaceSearchWidget(
...
                   expanded = false
                 },
                 modifier = Modifier.testTag("item-${place.place.id}"))
           }
         }
   }
 }
+
+/** function to convert a CustomPlace object to a Location object with only the fields we need */
+fun CustomPlace.toLocation(): Location {
```
Comment (apana03) [inline app/src/main/java/com/android/voyageur/ui/components/PlaceSearchWidget.kt:76]: This function should be in the CustomPlace class and used in the PlaceSearchWidget as this is an utility function that needs could be reused in other classes as well.

File: app/src/main/java/com/android/voyageur/ui/components/PlaceSearchWidget.kt (hunk 4 above)
Comment (apana03) [reply app/src/main/java/com/android/voyageur/ui/components/PlaceSearchWidget.kt:31]: For now it is fine, but we should look into it in order to make the code more modular. As this is not a very good practice :)

**Review by apana03 — APPROVED** [review]
Thank you for taking into account my previous comments and for finding a workaround for deserialization problem of the Place class. Just had a single comment regarding moving a function in another class, except that well done :)
